File: src/163MissingRanges/163MissingRanges.go
package main

import (
	"fmt"
	"strconv"
)

func findMissingRanges(nums []int, lower int, upper int) []string {
	n := len(nums)
	i := 0
	var diff int
	var result []string
	for i < n {
		if i == 0 {
			diff = nums[i] - lower
			if diff == 1 {
				result = append(result, strconv.Itoa(lower))
			} else if diff > 1 {
				result = append(result, strconv.Itoa(lower)+"->"+strconv.Itoa(nums[i]-1))
			}
		} else {
			diff = nums[i] - nums[i-1]
			if diff == 2 {
				result = append(result, strconv.Itoa(nums[i]-1))
			} else if diff > 2 {
				result = append(result, strconv.Itoa(nums[i-1]+1)+"->"+strconv.Itoa(nums[i]-1))
			}
		}
		i++
	}
	if i == 0 {
		if lower != upper {
			result = append(result, strconv.Itoa(lower)+"->"+strconv.Itoa(upper))
		} else {
			result = append(result, strconv.Itoa(lower))
		}
	} else {
		diff = upper - nums[i-1]
		if diff == 1 {
			result = append(result, strconv.Itoa(upper))
		} else if diff > 1 {
			result = append(result, strconv.Itoa(nums[i-1]+1)+"->"+strconv.Itoa(upper))
		}
	}

	return result
}

func main() {
	fmt.Printf("%v\n", findMissingRanges([]int{0, 1, 3, 50, 75}, 0, 99))
	fmt.Printf("%v\n", findMissingRanges([]int{0, 1, 3, 50, 75, 98}, 0, 99))
}

File: src/163MissingRanges/163MissingRanges.rs
impl Solution {
    pub fn find_missing_ranges(nums: Vec<i32>, lower: i32, upper: i32) -> Vec<String> {
        let mut result = vec![];
        if nums.len() == 0 {
            if upper == lower {
                result.push(format!("{}", lower));     
            } else {
                result.push(format!("{}->{}", lower, upper));
            }
        }else{
            for i in 0..nums.len() {
                if i == 0 {
                    if nums[i] as i64 - lower as i64 == 1 {
                        result.push(format!("{}", lower));     
                    } else if nums[i] as i64 - lower as i64 > 1 {
                        result.push(format!("{}->{}", lower, nums[i] - 1));
                    }
                }else{
                     if nums[i] as i64 - nums[i-1] as i64 == 2 {
                        result.push(format!("{}", nums[i-1]+1));     
                    } else if nums[i] as i64 - nums[i-1] as i64 > 2 {
                        result.push(format!("{}->{}", nums[i-1] + 1, nums[i] - 1));
                    }
                }

                if i==nums.len()-1 {
                    if upper as i64 - nums[i] as i64 == 1 {
                        result.push(format!("{}", upper));     
                    } else if upper as i64 - nums[i] as i64 > 1 {
                        result.push(format!("{}->{}", nums[i] + 1, upper));
                    }
                }
            }        
        }
        
        return result;
    }
}
